### PR TITLE
Add callback_url/acct information for Sidekiq PuSH workers Exception(alternative)

### PR DESCRIPTION
**Pull request by clworld**
This is alternative implementation of https://github.com/tootsuite/mastodon/pull/4281.
sidekiq_fixx

### CODE REVIEW

File: app/services/process_interaction_service.rb [hunk 1]
```diff
@@ -47,7 +47,7 @@ def call(envelope, target_account)
         reflect_unblock!(account, target_account)
       end
     end
-  rescue HTTP::Error, OStatus2::BadSalmonError, Mastodon::NotPermittedError
+  rescue HTTP::Error, OStatus2::BadSalmonError, Mastodon::NotPermittedError, Mastodon::WorkerError
```
Comment (clworld) [inline app/services/process_interaction_service.rb]: It's for deal with rspec error. WorkerError will be raised from inside of perform_async only when test.

File: app/workers/pubsubhubbub/delivery_worker.rb [hunk 2]
```diff
@@ -16,6 +16,8 @@ def perform(subscription_id, payload)
     @subscription = Subscription.find(subscription_id)
     @payload = payload
     process_delivery unless blocked_domain?
+  rescue => e
+    raise Mastodon::WorkerError, "Delivery failed for #{subscription&.callback_url}: #{e.class}: #{e.message}"
```
Comment (clworld) [inline app/workers/pubsubhubbub/delivery_worker.rb]: Wrap exception for adding callback_url to exception message.

File: app/workers/pubsubhubbub/subscribe_worker.rb [hunk 3]
```diff
@@ -22,5 +22,7 @@ def perform(account_id)
     account = Account.find(account_id)
     logger.debug "PuSH re-subscribing to #{account.acct}"
     ::SubscribeService.new.call(account)
+  rescue => e
+    raise Mastodon::WorkerError, "Subscribe failed for #{account&.acct}: #{e.class}: #{e.message}"
```
Comment (clworld) [inline app/workers/pubsubhubbub/subscribe_worker.rb]: Wrap exception for adding acct to exception message.

File: spec/workers/pubsubhubbub/delivery_worker_spec.rb [hunk 4]
```diff
@@ -10,7 +10,13 @@
...
   describe 'perform' do
     it 'raises when subscription does not exist' do
-      expect { subject.perform 123, payload }.to raise_error(ActiveRecord::RecordNotFound)
+      expect(
+        begin
+          subject.perform 123, payload
+        rescue Mastodon::WorkerError => e
+          e.cause
+        end
+      ).to be_an(ActiveRecord::RecordNotFound)
```
Comment (clworld) [inline spec/workers/pubsubhubbub/delivery_worker_spec.rb]: I think there is better way to write test to check e.cause... but I didn't found it..

File: app/lib/exceptions.rb [hunk 5]
```diff
@@ -5,6 +5,7 @@ class Error < StandardError; end
   class NotPermittedError < Error; end
   class ValidationError < Error; end
   class RaceConditionError < Error; end
+  class WorkerError < Error; end
```
Comment (clworld) [inline app/lib/exceptions.rb]: Add exception for wrap exception in Sidekiq worker. (wrapped exceptions are raised to Sidekiq)

**Comment by beatrix-bitrot** [comment]
I like it but I kind of want to know @Gargron thoughts on the matter since he recently changed some exception handling stuff

**Comment by Gargron** [comment]
I chose the other PR for simplicity.

**Pull request closed without merging**